### PR TITLE
feat: implement custom version of libp2p-tls

### PR DESCRIPTION
**Pull request by dignifiedquire**
- only ed25519 keys are supported
- no support for PeerIds that are based on protobufs
- remove support for RSA on the certificates

### CODE REVIEW

File: src/tls/certificate.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+use der::{asn1::OctetStringRef, Decode, Encode, Sequence};
+use x509_parser::prelude::*;
+
+use super::{Keypair, PeerId, PublicKey, Signature};
+
+/// The libp2p Public Key Extension is a X.509 extension
+/// with the Object Identier 1.3.6.1.4.1.53594.1.1,
+/// allocated by IANA to the libp2p project at Protocol Labs.
+const P2P_EXT_OID: [u64; 9] = [1, 3, 6, 1, 4, 1, 53594, 1, 1];
```
Comment (b5) [inline src/tls/certificate.rs:13]: can we get a TODO to not use this object identifier? It's tied to the protobuf you're trying to remove

File: src/tls/certificate.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        msg.extend(P2P_SIGNING_PREFIX);
+        msg.extend(certificate_keypair.public_key_der());
+
+        identity_keypair.sign(&msg)
+    };
+
+    let public_key = identity_keypair.public();
+    let signature = signature.to_bytes();
+    let key = SignedKey {
+        public_key: OctetStringRef::new(&public_key.as_bytes()[..]).unwrap(),
```
Comment (b5) [inline src/tls/certificate.rs:182]: IIUC this is the crucial change: dropping the keytype enum & protobuf serialization in favour of just the ED publickey directly, right?

**Review by b5 — APPROVED** [review]
ran tests locally, am satisfied that things still work. The comments / spec side of this is a mess, but I vote we clean that up later

File: src/tls/certificate.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/tls/certificate.rs:13]: we don't want to allocate our own at the IANA for now, so for now I don't think we should change this

**Pull request merged**